src/widgets/reflow.rs

```diff
@@ -72,155 +72,142 @@ where
-
-                                if symbol_width > remaining_width {
-                                    break;
-                                }
-                                remaining_width -= symbol_width;
-                                first_whitespace = unfinished_whitespaces.pop_front();
-                            }
-                            // In case all whitespaces have been exhausted
-                            if symbol_whitespace && first_whitespace.is_none() {
-                                // Prevent first whitespace to count towards next word
-                                continue;
-                            }
-                        }
+            let mut wrapped_lines = vec![];
+            let mut current_line = vec![];
+            let mut current_line_width = 0;
+            let mut pending_word = vec![];
+            let mut word_width = 0;
+            let mut pending_whitespace: VecDeque<StyledGrapheme> = VecDeque::new();
+            let mut whitespace_width = 0;
+            let mut non_whitespace_previous = false;
 
-                        // Append symbol to unfinished, partially processed word
-                        if symbol_whitespace {
-                            whitespace_width += symbol_width;
-                            unfinished_whitespaces.push_back(StyledGrapheme { symbol, style });
-                        } else {
-                            word_width += symbol_width;
-                            unfinished_word.push(StyledGrapheme { symbol, style });
-                        }
+            for StyledGrapheme { symbol, style } in line_symbols {
+                let is_whitespace =
+                    symbol == ZWSP || symbol.chars().all(char::is_whitespace) && symbol != NBSP;
+                let symbol_width = symbol.width() as u16;
+
+                // ignore symbols wider than max width
+                if symbol_width > self.max_line_width {
+                    continue;
+                }
 
-                        has_seen_non_whitespace = !symbol_whitespace;
+                // append finished word to current line
+                if non_whitespace_previous && is_whitespace
+                    || word_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && self.trim
+                    || whitespace_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && self.trim
+                    || word_width + whitespace_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && !self.trim
+                {
+                    if !current_line.is_empty() || !self.trim {
+                        current_line.extend(pending_whitespace.drain(..));
+                        current_line_width += whitespace_width;
                     }
```

Comment:
Hm, seems like it will also trigger in the last statement clause (the one with `!self.trim()`, but I definitely am having a hard time following the overall control flow, so I'm not even sure if it's possible...

Reply:
Yeah - that's kinda why I advocate for rewriting this - this might be a good first step in that (understanding what's there already).

If you wanted to really 100% nail the control flow of this, adding tests that incrementally hit every branch / repetition in the method would be the approach I'd take. I'd consider doing it starting from the paragraph in so that replacing this entirely would be possible.

Reply:
> adding tests that incrementally hit every branch / repetition

tbh, I'd save that for later to avoid exploding the scope of this PR if it's okay 😅